doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+own resources.
+
+The `init()` hook will trigger when an `AsyncEvent` is instantiated.
+ 
+It is important that before/after calls are unwound
+in the same order they are called. Otherwise an unrecoverable exception
+will be made.
+
+```js
+// AsyncEvent() is meant to be extended. Instantiating a new AsyncEvent() also
+// triggers init(). If triggerId is omitted then currentId() is used.
+const asyncEvent = new AsyncEvent(type[, triggerId]);
+
+// Call before() hooks.
+asyncEvent.emitBefore();
+
+// Call after() hooks.
+asyncEvent.emitAfter();
+
+// Call destroy() hooks.
+asyncEvent.emitDestroy();
+
+// Return the unique id assigned to the AsyncEvent instance.
+asyncEvent.asyncId();
+
+// Return the trigger id for the AsyncEvent instance.
+asyncEvent.triggerId();
+```
+
+#### `AsyncEvent(type[, triggerId])`
+
+* arguments
+  * `type` {string} the type of ascycn event
+  * `triggerId` {number} the id of the execution context that created this async
+    event
+* Returns {AsyncEvent} A reference to `asyncHook`.
+
+Example usage:
+
+```js
+class DBQuery extends AsyncEvent {
+  constructor(db) {
+    this.db = db;
+  }
+
+  getInfo(query, callback) {
+    this.db.get(query, (err, data) => {
+      this.emitBefore();
+      callback(err, data)
+      this.emitAfter();
+    });
+  }
+
+  close() {
+    this.db = null;
+    this.emitDestroy();
+  }
+}
+```
+
+#### `asyncEvent.emitBefore()`
+
+* Returns {undefined}
+
+Call all `before()` hooks and let them know a new asynchronous execution
+context is being entered. If nested calls to `emitBefore()` are made the stack
+of `id`s will be tracked and properly unwound.
+
+#### `asyncEvent.emitAfter()`
+
+* Returns {undefined}
+
+Call all `after()` hooks. If nested calls to `emitBefore()` were made then make
+sure the stack is unwound properly. Otherwise an error will be thrown.
+
+If the user's callback throws an exception then `emitAfter()` will
+automatically be called for all `id`'s on the stack if the error is handled by
+a domain or `'uncaughtException'` handler. So there is no need to guard against
+this.
+
+#### `asyncEvent.emitDestroy()`
+
+* Returns {undefined}
+
+Call all `destroy()` hooks. This should only ever be called once. An error will
+be thrown if it is called more than once. This **must** be manually called. If
+the resource is left to be collected by the GC then the `destroy()` hooks will
+never be called.
+
+#### `asyncEvent.asyncId()`
+
+* Returns {number} the unique `id` assigned to the resource. 
+
+Useful when used with `triggerIdScope()`.
+
+#### `asyncEvent.triggerId()`
+
+* Returns {number} the same `triggerId` that is passed to `init()` hooks.
+
+### Standalone JS API
+
+The following API can be used as an alternative to using `AsyncEvent()`, but it
+is left to the embedder to manually track values needed for all emitted events
+(e.g. `id` and `triggerId`). It mainly exists for use in Node.js core itself and
+it is highly recommended that embedders instead use `AsyncEvent`.
```

Comment:
why do we doc it? does it have a use-case? is there some rare thing it can be used for that the AsyncEvent API cannot be used for? Why don't we use the AsyncEvent API in node core? If its for efficiency, it seems worrying that we are suggesting user-land use an API that we won't use in node core.